# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
#lambda x: x +10
#regular named function
def add_ten(x):
    return x + 10
print(add_ten(4))

#lambda with variable
add_something = lambda x: x + 6

print(add_something(7))

#calling lambda withough  variable
print((lambda y: y *4)(8))

14
13
32


#### Saving to a Variable

In [6]:
x = lambda num: num +10
#def x(num):
# return num *10
print(x(10))
y = lambda name: print(name)
y('Boba Fett')

z= lambda num: int(num) + 80
print(z('40'))

20
Boba Fett
120


#### Multiple Inputs

In [11]:
# multiple inputs with no variable
print((lambda x, y, z: x* y *z)(3,5,8))

#multiple inputs with a variable name

x_test = lambda x, y, z: x* y + z
print(x_test(5, 8, 23))

string_test = lambda verb, adjective, noun: f" It is a {adjective} {noun} and I {verb} to the store."
print(string_test("roll","tasty","tree"))

120
63
 It is a tasty tree and I roll to the store.


#### Passing a Lambda into a Function

#### Returning a Lambda from a Function

#### If Statements within Lambdas

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in if they are less than 15, otherwise sqaure them. Assign the anonymous functions to avariable f'
</p>

In [16]:
anonymous_function = lambda num: num **3 if num < 15 else num**2

f = anonymous_function

print(f(15))
print(f(5))
print(f(20))

225
125
400


#### Syntax

#### Using Lambda's with Map

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [17]:
arr = [2 ,30 , 4, 15, 6, 17]

double_nums_subtract_one =list(map(lambda x: x*2 -1 , arr))
print(double_nums_subtract_one)

[3, 59, 7, 29, 11, 33]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

#### Using Lambda's with Filter()

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [19]:
from statistics import mean

nums = [2, 7, 4.2, 1.6, 9, 4.4, 5.9]


avge = list(filter(lambda a: True if a <= mean(nums) else False, nums))
print(avge)

[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [23]:
from functools import reduce
#reduce(functions, iterable)
list_1 = [2, 4, 6, 8, 10]

def add_nums(num1, num2):
    return num1 + num2
result_add = reduce(add_nums, list_1)

print(result_add)

#subtract a list of numbers

def subtract_nums(num1, num2):
    return num1 - num2
result_sub = reduce(subtract_nums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

#### Writing a Factorial Function

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### next() keyword
<p>Returns next element from the list, if not present prints the default value. If default value is not present, raises the StopIteration error.</p>

#### Creating a Generator Object
<p>The generator object will store an iterable on which we can call the next() method.</p>

<p>Use try and except to print graceful message instead of the StopIteration error:

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [24]:
def num_sq(stop,start,step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0,start = 3):
    print(i)

9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]



### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


### Exercise #4 <br>
<p>Create a generator function that individually returns each movie genre back from the list</p>




In [ ]:
genres = ["adventure", "drama", "horror", "comedy", "action", "romance"]

